hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/AbstractCGroupsHandler.java

```diff
@@ -0,0 +1,576 @@
+  }
+
+  /**
+   * This function is called when the administrator opted
+   * to use a pre-mounted cgroup controller.
+   * There are two options.
+   * 1. YARN hierarchy already exists. We verify, whether we have write access
+   * in this case.
+   * 2. YARN hierarchy does not exist, yet. We create it in this case. If cgroup v2 is used
+   * an additional step is required to update the cgroup.subtree_control file, see
+   * {@link CGroupsV2HandlerImpl#updateEnabledControllersInHierarchy}
+   *
+   * @param controller the controller being initialized
+   * @throws ResourceHandlerException yarn hierarchy cannot be created or
+   *                                  accessed for any reason
+   */
+  private void initializePreMountedCGroupController(CGroupController controller)
+      throws ResourceHandlerException {
+    // Check permissions to cgroup hierarchy and
+    // create YARN cgroup if it does not exist, yet
+    String controllerPath = getControllerPath(controller);
+
+    if (controllerPath == null) {
+      throw new ResourceHandlerException(
+          String.format("Controller %s not mounted."
+                  + " You either need to mount it with %s"
+                  + " or mount cgroups before launching Yarn",
+              controller.getName(), YarnConfiguration.
+                  NM_LINUX_CONTAINER_CGROUPS_MOUNT));
+    }
+
+    File rootHierarchy = new File(controllerPath);
+    File yarnHierarchy = new File(rootHierarchy, cGroupPrefix);
+    String subsystemName = controller.getName();
+
+    LOG.info("Initializing mounted controller " + controller.getName() + " " +
+        "at " + yarnHierarchy);
+
+    if (!rootHierarchy.exists()) {
+      throw new ResourceHandlerException(getErrorWithDetails(
+          "Cgroups mount point does not exist or not accessible",
+          subsystemName,
+          rootHierarchy.getAbsolutePath()
+      ));
+    } else if (!yarnHierarchy.exists()) {
+      LOG.info("Yarn control group does not exist. Creating " +
+          yarnHierarchy.getAbsolutePath());
+      try {
+        if (!yarnHierarchy.mkdir()) {
+          // Unexpected: we just checked that it was missing
+          throw new ResourceHandlerException(getErrorWithDetails(
+              "Unexpected: Cannot create yarn cgroup",
+              subsystemName,
+              yarnHierarchy.getAbsolutePath()
+          ));
+        }
+      } catch (SecurityException e) {
+        throw new ResourceHandlerException(getErrorWithDetails(
+            "No permissions to create yarn cgroup",
+            subsystemName,
+            yarnHierarchy.getAbsolutePath()
+        ), e);
+      }
+    } else if (!FileUtil.canWrite(yarnHierarchy)) {
+      throw new ResourceHandlerException(getErrorWithDetails(
+          "Yarn control group not writable",
+          subsystemName,
+          yarnHierarchy.getAbsolutePath()
+      ));
+    }
+
+    try {
+      updateEnabledControllersInHierarchy(yarnHierarchy);
```

Comment:
We have a chain of `ResourceHandler`(s) (`cpu`, `io`, ...) and the `.boostrap()` is called on each of them, where they call the `initializeCGroupController` with the actual controller (e.g. `cpu`) which then ends up here. But for cgroup v2 we initialise all valid cgroupv2 controller **at once** regardless of the configuration.

```
  private static void initializeConfiguredResourceHandlerChain(
      Configuration conf, Context nmContext)
      throws ResourceHandlerException {
    ArrayList<ResourceHandler> handlerList = new ArrayList<>();

    addHandlerIfNotNull(handlerList,
        initNetworkResourceHandler(conf));
    addHandlerIfNotNull(handlerList,
        initDiskResourceHandler(conf));
    addHandlerIfNotNull(handlerList,
        initMemoryResourceHandler(conf));
    addHandlerIfNotNull(handlerList,
        initCGroupsCpuResourceHandler(conf));
    addHandlerIfNotNull(handlerList, getNumaResourceHandler(conf, nmContext));
    addHandlersFromConfiguredResourcePlugins(handlerList, conf, nmContext);
    resourceHandlerChain = new ResourceHandlerChain(handlerList);
  }
```

Shouldn't the `updateEnabledControllersInHierarchy` be called only for a valid cgroupv2 controller and only enable that controller (if it is not already enabled)?

```
      if (controller.isInV2()) {
        updateEnabledControllersInHierarchy(yarnHierarchy, controller);
      }
```

Since this is an abstract class, the if shouldn't really be here but in the v2 implementation. This is just an example.

Reply:
I have some questions about this init phase, here is an example which can be used for discussion.

Enabled controllers in root:
```bash
tomi@tomi-pc: /sys/fs/cgroup
$cat cgroup.subtree_control
cpu memory pids
```


Available controllers and subtree_control for `mycgroup`, note user "tomi" has write permissions for `/sys/fs/cgroup/hadoop-yarn`:
```bash
tomi@tomi-pc: /sys/fs/cgroup
$cat mycgroup/cgroup.controllers
cpu memory pids
$cat mycgroup/cgroup.subtree_control
tomi@tomi-pc: /sys/fs/cgroup

tomi@tomi-pc: /sys/fs/cgroup
$ls -lh mycgroup/ | grep hadoop
drwxrwxr-x 4 root tomi 0 Apr 22 12:48 hadoop-yarn
```

Current state:
```bash
$tree mycgroup/
mycgroup/
├── cgroup.controllers
├── cgroup.events
├── cgroup.freeze
├── cgroup.kill
├── cgroup.max.depth
├── cgroup.max.descendants
├── cgroup.pressure
├── cgroup.procs
├── cgroup.stat
├── cgroup.subtree_control
├── cgroup.threads
├── cgroup.type
├── cpu.idle
├── cpu.max
├── cpu.max.burst
├── cpu.pressure
├── cpu.stat
├── cpu.stat.local
├── cpu.uclamp.max
├── cpu.uclamp.min
├── cpu.weight
├── cpu.weight.nice
├── hadoop-yarn
│   ├── cgroup.controllers
│   ├── cgroup.events
│   ├── cgroup.freeze
│   ├── cgroup.kill
│   ├── cgroup.max.depth
│   ├── cgroup.max.descendants
│   ├── cgroup.pressure
│   ├── cgroup.procs
│   ├── cgroup.stat
│   ├── cgroup.subtree_control
│   ├── cgroup.threads
│   ├── cgroup.type
│   ├── container-001
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   └── memory.pressure
│   ├── container-002
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   └── memory.pressure
│   ├── cpu.pressure
│   ├── cpu.stat
│   ├── cpu.stat.local
│   ├── io.pressure
│   ├── irq.pressure
│   └── memory.pressure
├── io.pressure
├── irq.pressure
├── memory.current
├── memory.events
├── memory.events.local
├── memory.high
├── memory.low
├── memory.max
├── memory.min
├── memory.numa_stat
├── memory.oom.group
├── memory.peak
├── memory.pressure
├── memory.reclaim
├── memory.stat
├── memory.swap.current
├── memory.swap.events
├── memory.swap.high
├── memory.swap.max
├── memory.swap.peak
├── memory.zswap.current
├── memory.zswap.max
├── memory.zswap.writeback
├── pids.current
├── pids.events
├── pids.max
└── pids.peak

4 directories, 103 files
```


Enabling controllers in `mycgroup`, note the user `tomi` doesn't have the permission for it:
```bash
tomi@tomi-pc: /sys/fs/cgroup
$echo "+cpu +memory +pids" > mycgroup/cgroup.subtree_control
bash: mycgroup/cgroup.subtree_control: Permission denied

tomi@tomi-pc: /sys/fs/cgroup
$sudo bash
[root@tomi-pc cgroup]# pwd
/sys/fs/cgroup
[root@tomi-pc cgroup]# echo "+cpu +memory +pids" > mycgroup/cgroup.subtree_control
[root@tomi-pc cgroup]#

tomi@tomi-pc: /sys/fs/cgroup
$cat mycgroup/cgroup.subtree_control
cpu memory pids
```

This means that we could check the `/sys/fs/cgroup/mycgroup/hadoop-yarn/cgroup.controllers` for available controllers,
but the parent path `/sys/fs/cgroup/mycgroup/` must have `cgroup.sub_controllers` enabled and YARN might not have the
permission to update those.


```bash
tomi@tomi-pc: /sys/fs/cgroup
$tree mycgroup/
mycgroup/
├── cgroup.controllers
├── cgroup.events
├── cgroup.freeze
├── cgroup.kill
├── cgroup.max.depth
├── cgroup.max.descendants
├── cgroup.pressure
├── cgroup.procs
├── cgroup.stat
├── cgroup.subtree_control
├── cgroup.threads
├── cgroup.type
├── cpu.idle
├── cpu.max
├── cpu.max.burst
├── cpu.pressure
├── cpu.stat
├── cpu.stat.local
├── cpu.uclamp.max
├── cpu.uclamp.min
├── cpu.weight
├── cpu.weight.nice
├── hadoop-yarn
│   ├── cgroup.controllers
│   ├── cgroup.events
│   ├── cgroup.freeze
│   ├── cgroup.kill
│   ├── cgroup.max.depth
│   ├── cgroup.max.descendants
│   ├── cgroup.pressure
│   ├── cgroup.procs
│   ├── cgroup.stat
│   ├── cgroup.subtree_control
│   ├── cgroup.threads
│   ├── cgroup.type
│   ├── container-001
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   └── memory.pressure
│   ├── container-002
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   └── memory.pressure
│   ├── cpu.idle
│   ├── cpu.max
│   ├── cpu.max.burst
│   ├── cpu.pressure
│   ├── cpu.stat
│   ├── cpu.stat.local
│   ├── cpu.uclamp.max
│   ├── cpu.uclamp.min
│   ├── cpu.weight
│   ├── cpu.weight.nice
│   ├── io.pressure
│   ├── irq.pressure
│   ├── memory.current
│   ├── memory.events
│   ├── memory.events.local
│   ├── memory.high
│   ├── memory.low
│   ├── memory.max
│   ├── memory.min
│   ├── memory.numa_stat
│   ├── memory.oom.group
│   ├── memory.peak
│   ├── memory.pressure
│   ├── memory.reclaim
│   ├── memory.stat
│   ├── memory.swap.current
│   ├── memory.swap.events
│   ├── memory.swap.high
│   ├── memory.swap.max
│   ├── memory.swap.peak
│   ├── memory.zswap.current
│   ├── memory.zswap.max
│   ├── memory.zswap.writeback
│   ├── pids.current
│   ├── pids.events
│   ├── pids.max
│   └── pids.peak
├── io.pressure
├── irq.pressure
├── memory.current
├── memory.events
├── memory.events.local
├── memory.high
├── memory.low
├── memory.max
├── memory.min
├── memory.numa_stat
├── memory.oom.group
├── memory.peak
├── memory.pressure
├── memory.reclaim
├── memory.stat
├── memory.swap.current
├── memory.swap.events
├── memory.swap.high
├── memory.swap.max
├── memory.swap.peak
├── memory.zswap.current
├── memory.zswap.max
├── memory.zswap.writeback
├── pids.current
├── pids.events
├── pids.max
└── pids.peak

4 directories, 134 files
```

User "tomi" (aka YARN) can enable controllers for the containers:
```bash
tomi@tomi-pc: /sys/fs/cgroup
$cat mycgroup/hadoop-yarn/cgroup.controllers
cpu memory pids
$cat mycgroup/hadoop-yarn/cgroup.subtree_control
tomi@tomi-pc: /sys/fs/cgroup
tomi@tomi-pc: /sys/fs/cgroup
$echo "+cpu +memory" > mycgroup/hadoop-yarn/cgroup.subtree_control
tomi@tomi-pc: /sys/fs/cgroup
$cat mycgroup/hadoop-yarn/cgroup.subtree_control
cpu memory
```

```bash
tomi@tomi-pc: /sys/fs/cgroup
$tree mycgroup/
mycgroup/
├── cgroup.controllers
├── cgroup.events
├── cgroup.freeze
├── cgroup.kill
├── cgroup.max.depth
├── cgroup.max.descendants
├── cgroup.pressure
├── cgroup.procs
├── cgroup.stat
├── cgroup.subtree_control
├── cgroup.threads
├── cgroup.type
├── cpu.idle
├── cpu.max
├── cpu.max.burst
├── cpu.pressure
├── cpu.stat
├── cpu.stat.local
├── cpu.uclamp.max
├── cpu.uclamp.min
├── cpu.weight
├── cpu.weight.nice
├── hadoop-yarn
│   ├── cgroup.controllers
│   ├── cgroup.events
│   ├── cgroup.freeze
│   ├── cgroup.kill
│   ├── cgroup.max.depth
│   ├── cgroup.max.descendants
│   ├── cgroup.pressure
│   ├── cgroup.procs
│   ├── cgroup.stat
│   ├── cgroup.subtree_control
│   ├── cgroup.threads
│   ├── cgroup.type
│   ├── container-001
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.idle
│   │   ├── cpu.max
│   │   ├── cpu.max.burst
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── cpu.uclamp.max
│   │   ├── cpu.uclamp.min
│   │   ├── cpu.weight
│   │   ├── cpu.weight.nice
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   ├── memory.current
│   │   ├── memory.events
│   │   ├── memory.events.local
│   │   ├── memory.high
│   │   ├── memory.low
│   │   ├── memory.max
│   │   ├── memory.min
│   │   ├── memory.numa_stat
│   │   ├── memory.oom.group
│   │   ├── memory.peak
│   │   ├── memory.pressure
│   │   ├── memory.reclaim
│   │   ├── memory.stat
│   │   ├── memory.swap.current
│   │   ├── memory.swap.events
│   │   ├── memory.swap.high
│   │   ├── memory.swap.max
│   │   ├── memory.swap.peak
│   │   ├── memory.zswap.current
│   │   ├── memory.zswap.max
│   │   └── memory.zswap.writeback
│   ├── container-002
│   │   ├── cgroup.controllers
│   │   ├── cgroup.events
│   │   ├── cgroup.freeze
│   │   ├── cgroup.kill
│   │   ├── cgroup.max.depth
│   │   ├── cgroup.max.descendants
│   │   ├── cgroup.pressure
│   │   ├── cgroup.procs
│   │   ├── cgroup.stat
│   │   ├── cgroup.subtree_control
│   │   ├── cgroup.threads
│   │   ├── cgroup.type
│   │   ├── cpu.idle
│   │   ├── cpu.max
│   │   ├── cpu.max.burst
│   │   ├── cpu.pressure
│   │   ├── cpu.stat
│   │   ├── cpu.stat.local
│   │   ├── cpu.uclamp.max
│   │   ├── cpu.uclamp.min
│   │   ├── cpu.weight
│   │   ├── cpu.weight.nice
│   │   ├── io.pressure
│   │   ├── irq.pressure
│   │   ├── memory.current
│   │   ├── memory.events
│   │   ├── memory.events.local
│   │   ├── memory.high
│   │   ├── memory.low
│   │   ├── memory.max
│   │   ├── memory.min
│   │   ├── memory.numa_stat
│   │   ├── memory.oom.group
│   │   ├── memory.peak
│   │   ├── memory.pressure
│   │   ├── memory.reclaim
│   │   ├── memory.stat
│   │   ├── memory.swap.current
│   │   ├── memory.swap.events
│   │   ├── memory.swap.high
│   │   ├── memory.swap.max
│   │   ├── memory.swap.peak
│   │   ├── memory.zswap.current
│   │   ├── memory.zswap.max
│   │   └── memory.zswap.writeback
│   ├── cpu.idle
│   ├── cpu.max
│   ├── cpu.max.burst
│   ├── cpu.pressure
│   ├── cpu.stat
│   ├── cpu.stat.local
│   ├── cpu.uclamp.max
│   ├── cpu.uclamp.min
│   ├── cpu.weight
│   ├── cpu.weight.nice
│   ├── io.pressure
│   ├── irq.pressure
│   ├── memory.current
│   ├── memory.events
│   ├── memory.events.local
│   ├── memory.high
│   ├── memory.low
│   ├── memory.max
│   ├── memory.min
│   ├── memory.numa_stat
│   ├── memory.oom.group
│   ├── memory.peak
│   ├── memory.pressure
│   ├── memory.reclaim
│   ├── memory.stat
│   ├── memory.swap.current
│   ├── memory.swap.events
│   ├── memory.swap.high
│   ├── memory.swap.max
│   ├── memory.swap.peak
│   ├── memory.zswap.current
│   ├── memory.zswap.max
│   ├── memory.zswap.writeback
│   ├── pids.current
│   ├── pids.events
│   ├── pids.max
│   └── pids.peak
├── io.pressure
├── irq.pressure
├── memory.current
├── memory.events
├── memory.events.local
├── memory.high
├── memory.low
├── memory.max
├── memory.min
├── memory.numa_stat
├── memory.oom.group
├── memory.peak
├── memory.pressure
├── memory.reclaim
├── memory.stat
├── memory.swap.current
├── memory.swap.events
├── memory.swap.high
├── memory.swap.max
├── memory.swap.peak
├── memory.zswap.current
├── memory.zswap.max
├── memory.zswap.writeback
├── pids.current
├── pids.events
├── pids.max
└── pids.peak

4 directories, 188 files
```

Reply:
Actually the `yarn.nodemanager.linux-container-executor.cgroups.mount-path` is the config for the base path (like /sys/fs/cgroup) and the `yarn.nodemanager.linux-container-executor.cgroups.hierarchy` would be the hierarchy under the base path (mycgroup in your example). I was trying to mimic the same behaviour we had on v1, which - based on the [doc](https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/NodeManagerCgroups.html) - requires either write access to create the directory from the cgroups.hierarchy config, or it requires a precreated directory with read/write access. I've edited the logic a bit: now we only add the controller specified in the init flow to the `cgroup.subtree_control` file, we do it regardless if it's present in the file already (has no effect if it is), and extended the error handling to say where exactly has the process failed. Additionally if the hierarchy directory is created by YARN we can assume it has read/write access on the mount_path. In this case the cgroup.subtree_control file is extended with the new controller in the parent directory as well, ensuring that the hierarchy will have it enabled.